pipeline-worker/src/pipeline-steps/summaries/index.ts

```
@@ -0,0 +1,230 @@
+    topicName,
+    summary,
+    usage,
+    cost,
+  });
+}
+
+/**
+ * Generate summaries for all topics in the tree.
+ *
+ * Each topic is processed individually in parallel to ensure all summaries
+ * are generated. The results are then aggregated with combined usage and cost.
+ *
+ * @param input - Input configuration with tree and LLM config
+ * @param apiKey - OpenAI API key
+ * @param options - Optional configuration (reportId, userId, enableWeave, weaveProjectName)
+ * @returns Result containing array of topic summaries with aggregated usage and cost, or an error
+ */
+export async function generateTopicSummaries(
+  input: SummariesInput,
+  apiKey: string,
+  options: ClusteringOptions = {},
+): Promise<Result<SummariesResult, ClusteringError>> {
+  const { tree, llm } = input;
+  const { reportId, userId, enableWeave, weaveProjectName } = options;
+
+  const reportLogger = getReportLogger("summaries", userId, reportId);
+
+  // Validate inputs
+  if (tree.length === 0) {
+    return failure(
+      new ClusteringError("Tree cannot be empty for summary generation"),
+    );
+  }
+
+  reportLogger.info(
+    { numTopics: tree.length },
+    "Starting individual topic summaries",
+  );
+
+  // Process each topic individually with controlled concurrency
+  // Using processBatchConcurrently to limit parallel API calls
+  const results = await processBatchConcurrently(
+    tree,
+    async ([topicName, topicData]) => {
+      return processTopicSummary(
+        topicName,
+        topicData,
+        llm,
+        apiKey,
+        reportId,
+        reportLogger,
+        { enableWeave, weaveProjectName },
+      );
+    },
+    MAX_CONCURRENT_SUMMARIES,
+  );
+
+  // Check for failures
+  const firstFailure = results.find((r) => r.tag === "failure");
+  if (firstFailure && firstFailure.tag === "failure") {
+    reportLogger.error(
+      { error: firstFailure.error },
+      "At least one topic summary failed",
+    );
+    return firstFailure;
```

Seems like this fails fast but the file header says that processing continues even with failures.